pkg/policy/api/rule.go

```diff
@@ -70,6 +83,63 @@ func NewRule() *Rule {
 	return &Rule{}
 }
 
+// MarshalJSON returns the JSON encoding of Rule r. We need to overwrite it to
+// enforce omitempty on the EndpointSelector nested structures.
+func (r *Rule) MarshalJSON() ([]byte, error) {
```

Comment:
Discussed offline

Reply:
Added a note to rework this in follow ups because it's less trivial than we thought. Using `reflect`, I might have a solution that doesn't require as many code changes and is easier on maintainability, but it still needs some work.

Reply:
I just noticed that this PR introduced custom JSON marshaling for `Rule`. This has some implications for https://github.com/cilium/cilium/pull/11607. Specifically, one of the reasons I forked `controller-tools` is to remove the constraint of a type implementing custom JSON marshaling cannot generate a validation schema. In the upstream, https://github.com/kubernetes-sigs/controller-tools/pull/427 is responsible for this. In the fork, I've reverted that support. 

My question is why do we need this? I [asked on K8s Slack](https://kubernetes.slack.com/archives/CAR30FCJZ/p1596152053329900?thread_ts=1596146930.326300&cid=CAR30FCJZ) about opting out of this feature in controller-tools.

Reply:
We need this to properly implement the `omitempty` tags of `NodeSelector` and `EndpointSelector`. Because they are not pointers, without the custom marshal, marshalling then unmarshalling would create a new field where none existed. That is, when marshalling it would check `r.EndpointSelector` (always `!= nil`) instead of `r.EndpointSelector.LabelSelector` and thus always create the corresponding JSON entry.

You can easily reproduce this by removing the custom marshalling and running unit tests on the package. One of the tests checks `json.Unmarshal(json.Marhsal(rule)) == rule`.